### PR TITLE
Add Python bindings for EOS and TOV solution

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Gets some of the technical difficulties out of the way for adding more Python bindings, and adds bindings for the polytropic EOS and the TOV solution.

- [x] The first commit is factored into #1763 

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/PythonBindings/CharmCompatibility.cpp [hunk 1]
```diff
@@ -66,3 +67,24 @@ std::string get_library_versions() noexcept {
...
+
+/*
+ * PUP::able support
+ *
+ * Here we provide dummy-implementations of functions in Charm++'s `pup.h` so
+ * that we can write Python wrappers for classes that derive from `PUP::able`.
+ * These functions are not intended to be called.
+ */
+
+PUP::able::~able() {}
```
Comment (nilsdeppe) [inline src/PythonBindings/CharmCompatibility.cpp]: Should we throw exceptions in the functions that shouldn't be called? That way if someone calls them unintentionally we can give some sort of error

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Python/Tov.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+namespace bp = boost::python;
+
+namespace gr {
+namespace Solutions {
+namespace py_bindings {
+
+void bind_tov() {
+  bp::class_<TovSolution, boost::noncopyable>(
+      "Tov",
+      // Only allowing a polytropic EOS here for now since boost::python has
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Python/Tov.cpp]: I haven't tried using virtual base classes with bp. What troubles have you run into? Are they so bad we should consider switching to bypind11 in the future?

Edit:
There's some info  here: https://www.boost.org/doc/libs/1_61_0/libs/python/doc/html/tutorial/tutorial/exposing.html#tutorial.exposing.inheritance

**Review by nilsdeppe — COMMENTED** [review]
Overall looks good. A couple things I'd like some discussion on before approving :)

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Python/Tov.cpp (hunk 2 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Python/Tov.cpp]: I got virtual base classes working now, details below. There might be another reason to switch to pybind11 though: Looks like it deals with numpy interoperability quite well, e.g. provides functionality to vectorize a C++ function that takes and returns `double`s into a Python function that takes and returns Numpy arrays, and a way to expose our DataStructures to Numpy as buffers. (Edit: https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html)

Re virtual base classes: The tutorial you mentioned is the one I followed to get the virtual base classes working before. The issue I ran into turned out to be only tangentially related to virtual base classes. Apparently the type of the `equation_of_state` argument must be correctly specified as const-reference when it's a virtual base class, but doesn't need to be when it's a non-virtual subclass. Mostly I find that bp compiler errors are quite unhelpful, but it seems to be working OK.

**Comment by nilsvu** [comment]
@nilsdeppe please take a look at the fixups.

Perhaps @fmahebert could you do the second review since you are editing TOV star code?

**Comment by fmahebert** [comment]
Sure, I'm happy to take a look, but I know very little about Python bindings so mine will be a non-expert review...

**Review by nilsdeppe — COMMENTED** [review]
LTGM. I'm happy for you to squash

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/CMakeLists.txt [hunk 3]
```diff
@@ -17,3 +17,5 @@ add_test_library(
   "${LIBRARY_SOURCES}"
   "Test_Pypp;GeneralRelativitySolutions;GeneralizedHarmonic;GeneralRelativity;Utilities"
   )
+
```
Comment (moxcodes) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/CMakeLists.txt:20]: Probably best to double-check that this compiles without python bindings enabled.

File: cmake/SpectreSetupPython.cmake [hunk 4]
```diff
@@ -11,6 +11,7 @@ set(SPECTRE_LINK_PYBINDINGS
   -Wl,--whole-archive
   PUBLIC PyBindings
   PUBLIC ${HDF5_LIBRARIES}
+  PUBLIC ${GSL_LIBRARIES}
```
Comment (moxcodes) [inline cmake/SpectreSetupPython.cmake]: @nilsleiffischer says this doesn't need to be in this PR.

File: src/PythonBindings/CharmCompatibility.cpp [hunk 5]
```diff
@@ -76,15 +77,25 @@ std::string get_paths() noexcept { return "Not supported in python."; }
...
-PUP::able::~able() {}
-
-void PUP::able::pup(PUP::er& /*p*/) {}
+PUP::able::~able() {
+  // Destructor will be called in Python, but doesn't need to do anything
+}
 
-void PUP::able::PUP_ID::setName(const char* /*name*/) {}
+void PUP::able::pup(PUP::er& /*p*/) {
+  throw std::runtime_error{
```
Comment (moxcodes) [inline src/PythonBindings/CharmCompatibility.cpp:95]: probably useful to give a code comment about what this ends up being in the corresponding python.

File: tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Python/Test_PolytropicFluid.py [hunk 6]
```diff
@@ -10,7 +10,8 @@ class TestPolytropicFluid(unittest.TestCase):
     def test_creation(self):
         # This class currently exposes no functionality, so we can't test
         # anything but its base class and that it can be instantiated.
-        eos = spectre_eos.RelativisticPolytropicFluid(100.0, 2.0)
+        eos = spectre_eos.RelativisticPolytropicFluid(
+            polytropic_constant=100., polytropic_exponent=2.0)
```
Comment (moxcodes) [inline tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Python/Test_PolytropicFluid.py]: `.` vs `.0`

File: src/PythonBindings/CharmCompatibility.cpp [hunk 7]
```diff
@@ -77,10 +79,18 @@ std::string get_paths() noexcept { return "Not supported in python."; }
  * These functions are not intended to be called.
  */
 
-PUP::able::~able() {
+// clang-tidy: can't use `= default` to define a trivial destructor
+PUP::able::~able() {  // NOLINT
```
Comment (moxcodes) [inline src/PythonBindings/CharmCompatibility.cpp]: try doing what clang-tidy say and see what breaks?

**Review by moxcodes — COMMENTED** [review]
group code review with @nilsleiffischer and @fmahebert (primary)

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/CMakeLists.txt (hunk 3 above)
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/CMakeLists.txt:20]: The `spectre_add_python_test` function checks if `BUILD_PYTHON_BINDINGS` is enabled, so this is fine. See e.g. `tests/Informer/CMakeLists.txt`.

File: cmake/SpectreSetupPython.cmake (hunk 4 above)
Comment (nilsvu) [reply cmake/SpectreSetupPython.cmake]: Dropped the commit. I experimented with cubic spline interpolation which requires GSL, but that's unrelated to this PR.

File: src/PythonBindings/CharmCompatibility.cpp (hunk 7 above)
Comment (nilsvu) [reply src/PythonBindings/CharmCompatibility.cpp]: it didn't expect this to work but it did 😄 thanks 👍

**Comment by nilsvu** [comment]
@fmahebert could you take a look at the fixups? Thanks!

**Comment by fmahebert** [comment]
I am happy with the fixups, please squash 👍

**Comment by nilsvu** [comment]
@nilsdeppe do you want to check the fixups as well?

Canceled the Travis build to speed up others

**Comment by nilsdeppe** [comment]
LGTM :+1: I'm happy for you to squash

**Comment by nilsdeppe** [comment]
Why did you force-push? Please leave a comment if you push over something that's been reviewed so that reviewers know what changed.

**Comment by nilsvu** [comment]
Right, I rebased on #1761 that was recently merged

**Review by fmahebert — APPROVED** [review]

**Comment by kidder** [comment]
@nilsdeppe @moxcodes please look at

**Review by moxcodes — APPROVED** [review]
Looks good.

**Review by nilsdeppe — APPROVED** [review]

**Comment by kidder** [comment]
ignoring travis

**Pull request merged**